// zephyr-agent@0.0.53/dist/lib/transformers/ze-build-snapshot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSnapshot = createSnapshot;
const zephyr_edge_contract_1 = require("zephyr-edge-contract");
const ze_basehref_handler_1 = require("./ze-basehref-handler");
const errors_1 = require("../errors");
async function createSnapshot(zephyr_engine, { mfConfig, assets }) {
    const buildId = await zephyr_engine.build_id;
    if (!buildId) {
        await zephyr_engine.build_id;
        throw new errors_1.ZephyrError(errors_1.ZeErrors.ERR_DEPLOY_LOCAL_BUILD, {
            message: 'Cannot create snapshot before getting buildId',
        });
    }
    const options = {
        git_branch: zephyr_engine.gitProperties.git.branch,
        buildId,
        username: (await zephyr_engine.application_configuration).username,
        email: (await zephyr_engine.application_configuration).email,
        applicationProperties: zephyr_engine.applicationProperties,
        edge_url: (await zephyr_engine.application_configuration).EDGE_URL,
        gitProperties: zephyr_engine.gitProperties,
        mfConfig: mfConfig,
    };
    const version_postfix = zephyr_engine.env.isCI
        ? `${options.git_branch}.${options.buildId}`
        : `${options.username}.${options.buildId}`;
    const basedAssets = (0, ze_basehref_handler_1.applyBaseHrefToAssets)(assets, zephyr_engine.buildProperties.baseHref);
    return {
        // ZeApplicationProperties
        application_uid: (0, zephyr_edge_contract_1.createApplicationUid)(options.applicationProperties),
        version: `${options.applicationProperties.version}-${version_postfix}`,
        // ZeApplicationProperties + buildId + ZeApplicationProperties.username
        snapshot_id: (0, zephyr_edge_contract_1.flatCreateSnapshotId)(Object.assign({}, options.applicationProperties, {
            buildId: options.buildId,
            username: options.username,
        })),
        domain: options.edge_url,
        uid: {
            build: options.buildId,
            app_name: options.applicationProperties.name,
            repo: options.applicationProperties.project,
            org: options.applicationProperties.org,
        },
        git: options.gitProperties.git,
        creator: {
            name: options.username,
            email: options.email,
        },
        createdAt: Date.now(),
        mfConfig: options.mfConfig,
        assets: Object.keys(basedAssets).reduce((memo, hash) => {
            const asset = basedAssets[hash];
            const { path, extname, size } = asset;
            memo[asset.path] = { path, extname, hash: asset.hash, size };
            return memo;
        }, {}),
    };
}
