// zephyr-agent@0.0.53/dist/lib/transformers/ze-build-dash-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.zeBuildDashData = zeBuildDashData;
const errors_1 = require("../errors");
async function zeBuildDashData(zephyr_engine) {
    var _a, _b;
    const snapshotId = await zephyr_engine.snapshotId;
    const buildId = await zephyr_engine.build_id;
    if (!snapshotId) {
        throw new errors_1.ZephyrError(errors_1.ZeErrors.ERR_SNAPSHOT_ID_NOT_FOUND);
    }
    if (!buildId) {
        throw new errors_1.ZephyrError(errors_1.ZeErrors.ERR_GET_BUILD_ID);
    }
    const application_uid = zephyr_engine.application_uid;
    const isCI = zephyr_engine.env.isCI;
    const git = zephyr_engine.gitProperties.git;
    const app = zephyr_engine.applicationProperties;
    const name = zephyr_engine.applicationProperties.name;
    const { EDGE_URL: edge_url, username, DELIMITER: delimiter, } = await zephyr_engine.application_configuration;
    const to_raw = _recordToRawDependency;
    return {
        id: application_uid,
        name,
        environment: '',
        edge: { url: edge_url, delimiter },
        app: Object.assign({}, app, {
            buildId,
        }),
        version: snapshotId,
        git,
        context: { isCI, username },
        dependencies: to_raw(zephyr_engine.npmProperties.dependencies),
        devDependencies: to_raw(zephyr_engine.npmProperties.devDependencies),
        optionalDependencies: to_raw(zephyr_engine.npmProperties.optionalDependencies),
        peerDependencies: to_raw(zephyr_engine.npmProperties.peerDependencies),
        overrides: [],
        consumes: [],
        modules: [],
        remotes: (_b = (_a = zephyr_engine.federated_dependencies) === null || _a === void 0 ? void 0 : _a.map((r) => r.name)) !== null && _b !== void 0 ? _b : [],
        tags: [],
        project: '',
        metadata: {},
        default: false,
        remote: 'remoteEntry.js',
        type: 'app',
    };
}
function _recordToRawDependency(record) {
    if (!record)
        return [];
    return Object.entries(record).map(([name, version]) => ({ name, version }));
}
