// vite-plugin-zephyr@0.0.53/dist/lib/internal/extract/load_static_entries.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.load_static_entries = load_static_entries;
const node_path_1 = require("node:path");
const node_fs_1 = require("node:fs");
const vite_1 = require("vite");
const node_util_1 = require("node:util");
async function load_static_entries(props) {
    const { root } = props;
    const publicAssets = [];
    const root_dist_dir = (0, node_path_1.resolve)(root, props.outDir);
    const loadDir = async (destDir) => {
        for (const file of (0, node_fs_1.readdirSync)(destDir)) {
            const destFile = (0, node_path_1.resolve)(destDir, file);
            const stat = (0, node_fs_1.statSync)(destFile);
            if (stat.isDirectory()) {
                await loadDir(destFile);
                continue;
            }
            const fileName = (0, vite_1.normalizePath)((0, node_path_1.relative)(root_dist_dir, destFile));
            publicAssets.push({
                fileName,
                name: file,
                names: [file],
                needsCodeReference: false,
                source: await (0, node_util_1.promisify)(node_fs_1.readFile)(destFile),
                type: 'asset',
                originalFileName: file,
                originalFileNames: [file],
            });
        }
    };
    await loadDir(root_dist_dir);
    return publicAssets;
}
