// zephyr-agent@0.0.53/dist/lib/logging/ze-log-event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logFn = void 0;
exports.formatLogMsg = formatLogMsg;
exports.logger = logger;
const zephyr_edge_contract_1 = require("zephyr-edge-contract");
const get_application_configuration_1 = require("../edge-requests/get-application-configuration");
const errors_1 = require("../errors");
const http_request_1 = require("../http/http-request");
const token_1 = require("../node-persist/token");
const debug_1 = require("./debug");
const logFn = (level, msg) => {
    const formatted = formatLogMsg(msg, level);
    switch (level) {
        case 'warn':
            console.warn(formatted);
            break;
        case 'debug':
            console.debug(formatted);
            break;
        case 'error':
            console.error(formatted);
            break;
        default:
            console.log(formatted);
    }
};
exports.logFn = logFn;
function toLevelPrefix(level) {
    switch (level) {
        case 'warn':
            return debug_1.brightYellowBgName;
        case 'debug':
            return debug_1.brightGreenBgName;
        case 'error':
            return debug_1.brightRedBgName;
        default:
            return debug_1.brightBlueBgName;
    }
}
function formatLogMsg(msg, level = 'info') {
    return String(msg)
        .split('\n')
        .map((m) => `${toLevelPrefix(level)}  ${m.trimEnd()}`)
        .join('\n');
}
function logger(props) {
    const { application_uid, buildId, git } = props;
    // lazy loads the application configuration and token
    const loadLogData = (0, zephyr_edge_contract_1.PromiseLazyLoad)(() => {
        return Promise.all([(0, get_application_configuration_1.getApplicationConfiguration)({ application_uid }), (0, token_1.getToken)()]);
    });
    const url = new URL(zephyr_edge_contract_1.ze_api_gateway.logs, (0, zephyr_edge_contract_1.ZE_API_ENDPOINT)());
    return function logEvent(...logs) {
        // Prints logs to the console as fast as possible
        for (const log of logs) {
            if (!log.level && !log.action) {
                throw new errors_1.ZephyrError(errors_1.ZeErrors.ERR_UNKNOWN, {
                    message: 'Log level and action are required',
                });
            }
            (0, exports.logFn)(log.level, log.message);
        }
        // Then attempt to upload logs,
        loadLogData()
            .then(([config, token]) => void (0, http_request_1.makeRequest)(url, {
            method: 'POST',
            headers: {
                Authorization: `Bearer ${token}`,
                'Content-Type': 'application/json',
            },
        }, JSON.stringify(logs
            // some logs are empty to give newline effect in terminal
            .filter((l) => l.message.length && !l.ignore)
            .map((log) => ({
            application_uid: application_uid,
            userId: config.user_uuid,
            username: config.username,
            zeBuildId: buildId,
            logLevel: log.level,
            actionType: log.action,
            git: git,
            message: (0, zephyr_edge_contract_1.stripAnsi)(log.message.trim()),
            createdAt: Date.now(),
        })))))
            // This is ok to fail silently
            .catch(() => void 0);
    };
}
