// zephyr-agent@0.0.53/dist/lib/node-persist/storage-keys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageKeys = exports.ZE_SESSION_LOCK = exports.ZE_PATH = void 0;
const tslib_1 = require("tslib");
const path = tslib_1.__importStar(require("node:path"));
const fs = tslib_1.__importStar(require("node:fs"));
const os = tslib_1.__importStar(require("node:os"));
exports.ZE_PATH = path.resolve(os.homedir(), '.zephyr');
exports.ZE_SESSION_LOCK = path.resolve(exports.ZE_PATH, 'session');
try {
    // Ensures that the directory exists and lockfile is writable
    fs.mkdirSync(exports.ZE_PATH, { recursive: true });
}
catch (error) {
    console.error('error', `Could not create ~/.zephyr directory. Please check your permissions: ${error}`);
}
var StorageKeys;
(function (StorageKeys) {
    StorageKeys["ze_app_partial_asset_map"] = "ze_app_partial_asset_map";
    StorageKeys["ze_app_config_token"] = "ze-application-configuration";
    StorageKeys["ze_auth_token"] = "ze-auth-token";
    StorageKeys["ze_secret_token"] = "ZE_SECRET_TOKEN";
    StorageKeys["ze_fs_cache"] = "ze-fs-cache";
    StorageKeys["ze_hash_cache"] = "ze-hash-cache";
    StorageKeys["ze_app_deploy_result"] = "ze-app-deploy-result";
})(StorageKeys || (exports.StorageKeys = StorageKeys = {}));
