// zephyr-edge-contract@0.0.53/dist/lib/api-contract-negotiation/get-api-contract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ze_api_gateway = exports.ZE_IS_PREVIEW = exports.ZE_API_ENDPOINT_HOST = exports.ZE_API_ENDPOINT = exports.ZEPHYR_API_ENDPOINT = void 0;
/* istanbul ignore file */
const ZEPHYR_API_ENDPOINT = () => { var _a; return ((_a = process.env['ZE_API']) === null || _a === void 0 ? void 0 : _a.length) ? process.env['ZE_API'] : 'https://api.zephyr-cloud.io'; };
exports.ZEPHYR_API_ENDPOINT = ZEPHYR_API_ENDPOINT;
const ZE_API_ENDPOINT = () => {
    var _a;
    return ((_a = process.env['ZE_API_GATE']) === null || _a === void 0 ? void 0 : _a.length)
        ? process.env['ZE_API_GATE']
        : 'https://zeapi.zephyrcloud.app';
};
exports.ZE_API_ENDPOINT = ZE_API_ENDPOINT;
const ZE_API_ENDPOINT_HOST = () => new URL((0, exports.ZE_API_ENDPOINT)()).host;
exports.ZE_API_ENDPOINT_HOST = ZE_API_ENDPOINT_HOST;
const ZE_IS_PREVIEW = () => process.env['ZE_IS_PREVIEW'] === 'true';
exports.ZE_IS_PREVIEW = ZE_IS_PREVIEW;
exports.ze_api_gateway = {
    logs: '/logs',
    build_stats: '/build-stats',
    auth_link: '/auth-link',
    resolve: '/resolve',
    application_config: '/application-config',
};
