// zephyr-edge-contract@0.0.53/dist/lib/promise/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.forEachLimit = void 0;
exports.PromiseWithResolvers = PromiseWithResolvers;
exports.deferred = deferred;
exports.PromiseLazyLoad = PromiseLazyLoad;
exports.PromiseTuple = PromiseTuple;
exports.isSuccessTuple = isSuccessTuple;
var forEachLimit_1 = require("./forEachLimit");
Object.defineProperty(exports, "forEachLimit", { enumerable: true, get: function () { return forEachLimit_1.forEachLimit; } });
/**
 * ES 2023 Promise.withResolvers() polyfill
 *
 * @link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers
 */
function PromiseWithResolvers() {
    let resolve;
    let reject;
    const promise = new Promise((res, rej) => {
        resolve = res;
        reject = rej;
    });
    return { promise, resolve, reject };
}
/**
 * Creates a deferred promise.
 *
 * @returns A tuple containing the promise, resolve and reject functions in array
 */
function deferred() {
    let resolve;
    let reject;
    const promise = new Promise((res, rej) => {
        resolve = res;
        reject = rej;
    });
    return [promise, resolve, reject];
}
/**
 * Lazy loads a promise.
 *
 * @link https://github.com/sindresorhus/p-lazy
 */
function PromiseLazyLoad(promise) {
    let _data;
    return (params) => {
        if (!_data) {
            _data = promise(params);
        }
        return _data;
    };
}
/** Creates a `[error, value]` tuple value from a promise. */
async function PromiseTuple(maybePromise) {
    try {
        return [null, await maybePromise];
    }
    catch (err) {
        return [err, undefined];
    }
}
function isSuccessTuple(tuple) {
    return tuple[0] === null;
}
