// zephyr-agent@0.0.53/dist/lib/edge-actions/ze-upload-assets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.zeUploadAssets = zeUploadAssets;
const zephyr_edge_contract_1 = require("zephyr-edge-contract");
const upload_file_1 = require("../http/upload-file");
const logging_1 = require("../logging");
const picocolor_1 = require("../logging/picocolor");
const CLOUDFLARE_BATCH_SIZE = 6;
async function zeUploadAssets(zephyr_engine, { missingAssets, assetsMap }) {
    const count = Object.keys(assetsMap).length;
    const logger = await zephyr_engine.logger;
    const appConfig = await zephyr_engine.application_configuration;
    if (missingAssets.length === 0) {
        logger({
            level: 'info',
            action: 'snapshot:assets:upload:empty',
            message: `No assets to upload, ${(0, picocolor_1.white)('skipping')}...`,
        });
        return true;
    }
    const start = Date.now();
    let totalSize = 0;
    // If the target is iOS or Android, we upload the assets in a 6 request batch to avoid cloudflare worker requests limit
    // Reference: https://developers.cloudflare.com/workers/platform/limits/#simultaneous-open-connections:~:text=Once%20an%20invocation%20has%20six%20connections%20open%2C%20it%20can%20still%20attempt%20to%20open%20additional%20connections.
    if (zephyr_engine.env.target !== 'ios' && zephyr_engine.env.target !== 'android') {
        await Promise.all(missingAssets.map(upload_missing_asset));
    }
    else {
        logging_1.ze_log.upload("The target platform is 'ios' and 'android' so we are switching to batch upload.");
        await (0, zephyr_edge_contract_1.forEachLimit)(missingAssets.map((asset) => () => upload_missing_asset(asset)), CLOUDFLARE_BATCH_SIZE);
    }
    logger({
        level: 'info',
        action: 'snapshot:assets:upload:done',
        message: (0, picocolor_1.white)(`(${(0, picocolor_1.whiteBright)(missingAssets.length.toString())}/${(0, picocolor_1.white)(count.toString())} assets uploaded in ${(0, picocolor_1.whiteBright)((Date.now() - start).toString())}ms, ${(0, picocolor_1.whiteBright)(totalSize.toFixed(2))}kb)`),
    });
    return true;
    async function upload_missing_asset(asset) {
        var _a;
        const start = Date.now();
        const assetWithBuffer = assetsMap[asset.hash];
        const assetSize = ((_a = assetWithBuffer === null || assetWithBuffer === void 0 ? void 0 : assetWithBuffer.buffer) === null || _a === void 0 ? void 0 : _a.length) / 1024;
        await (0, upload_file_1.uploadFile)({
            hash: asset.hash,
            asset: assetWithBuffer,
        }, appConfig);
        const fileUploaded = Date.now() - start;
        totalSize += assetSize;
        logging_1.ze_log.upload(`file ${asset.path} uploaded in ${fileUploaded}ms (${assetSize.toFixed(2)}kb)`);
    }
}
