// zephyr-agent@0.0.53/dist/lib/http/upload-envs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uploadEnvs = uploadEnvs;
const get_application_configuration_1 = require("../edge-requests/get-application-configuration");
const errors_1 = require("../errors");
const logging_1 = require("../logging");
const http_request_1 = require("./http-request");
async function uploadEnvs({ body, application_uid, logEvent, }) {
    logging_1.ze_log.upload(`Uploading envs to Zephyr, for ${application_uid}`);
    const { EDGE_URL, jwt } = await (0, get_application_configuration_1.getApplicationConfiguration)({
        application_uid,
    });
    const json = JSON.stringify(body);
    const options = {
        method: 'POST',
        headers: {
            'Content-Length': Buffer.byteLength(json).toString(),
            'Content-Type': 'application/json; charset=utf-8',
            can_write_jwt: jwt,
        },
    };
    const url = new URL('/upload', EDGE_URL);
    url.searchParams.append('type', 'envs');
    const [ok, cause, data] = await (0, http_request_1.makeRequest)(url, options, JSON.stringify(body));
    if (!ok || !data) {
        logEvent({
            level: 'error',
            action: 'deploy:edge:failed',
            message: 'failed deploying local build to edge',
        });
        throw new errors_1.ZephyrError(errors_1.ZeErrors.ERR_FAILED_UPLOAD, { type: 'envs', cause: cause });
    }
}
